### PR TITLE
Removal of PlatformAbstractions package

### LINKED ISSUE #18438: Removal of Microsoft.DotNet.PlatformAbstractions Package
## Removal of Microsoft.DotNet.PlatformAbstractions Package

We will stop building new versions of [Microsoft.DotNet.PlatformAbstractions](https://www.nuget.org/packages/Microsoft.DotNet.PlatformAbstractions/) in .NET 5

### Version introduced

.NET 5.0 Preview 6

### Old behavior

New versions of the `Microsoft.DotNet.PlatformAbstractions` library were produced alongside new versions of .NET Core.

### New behavior

New functionality and major versions of this library will no longer be produced.

### Reason for change

As part of the [ongoing repository consolidation effort](https://github.com/dotnet/announcements/issues/119) in .NET 5, we have been examining the libraries that were previously in [dotnet/core-setup](https://github.com/dotnet/core-setup) and deciding on their long-term strategy/goals.

The [Microsoft.DotNet.PlatformAbstractions](https://www.nuget.org/packages/Microsoft.DotNet.PlatformAbstractions/) library has considerable overlap with APIs that are already established in the System.* namespaces. Also, some PlatformAbstractions APIs were not designed with the same level of scrutiny and long-term supportability that the rest of the System.* APIs are designed. For example, PlatformAbstractions uses an enum named `Platform` to describe the current Operating System Platform. This design was explicitly rejected when designing `RuntimeInformation.IsOSPlatform(OSPlatform)` to allow for new platforms and to allow the most flexibility in the future.

To resolve this situation, we have ensured that the scenarios enabled by `Microsoft.DotNet.PlatformAbstractions` are now possible without it and plan to stop building new versions of `Microsoft.DotNet.PlatformAbstractions` in .NET 5. The existing versions will continue to work, even in .NET 5, and be serviced along with previous versions of .NET Core. But new functionality won't be added to the library. New functionality will be added through other libraries and APIs.

### Recommended action

You can continue to use older versions of the library if they meet your requirements.

When the older version no longer meets your requirements, you can replace usages of the PlatformAbstractions APIs with the recommended API in the following table.

| PlatformAbstractions API                   | Recommended Replacement                                       |
|--------------------------------------------|---------------------------------------------------|
| ApplicationEnvironment.ApplicationBasePath | [AppContext.BaseDirectory][1]                     |
| HashCodeCombiner                           | [System.HashCode][2]                              |
| RuntimeEnvironment.GetRuntimeIdentifier()  | [RuntimeInformation.RuntimeIdentifier][3]         |
| RuntimeEnvironment.OperatingSystemPlatform | [RuntimeInformation.IsOSPlatform(OSPlatform)][4]  |
| RuntimeEnvironment.RuntimeArchitecture     | [RuntimeInformation.ProcessArchitecture][5]       |
| RuntimeEnvironment.OperatingSystem         | [RuntimeInformation.OSDescription][6] ** See Note |
| RuntimeEnvironment.OperatingSystemVersion  | [RuntimeInformation.OSDescription][6] and [Environment.OSVersion.Version][7] ** See Note |

[1]: https://docs.microsoft.com/dotnet/api/system.appcontext.basedirectory
[2]: https://docs.microsoft.com/dotnet/api/system.hashcode
[3]: https://github.com/dotnet/runtime/pull/34206
[4]: https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.runtimeinformation.isosplatform
[5]: https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.runtimeinformation.processarchitecture
[6]: https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.runtimeinformation.osdescription
[7]: https://docs.microsoft.com/dotnet/api/system.environment.osversion

#### Note on OperatingSystem and OperatingSystemVersion

The majority of use cases for these APIs are for display purposes: displaying to a user, logging, telemetry, etc. Making runtime decisions on an operating system version (especially one that is of type `string`) is not a recommended practice. We have made enhancements to `Environment.OSVersion.Version` to return the correct version for Windows and macOS Operating Systems. However, for most Unix distributions, what is considered to be the "OS Version" is not as straight-forward. Do you want the Linux kernel version? The distro version? etc. For most Unix platforms, `Environment.OSVersion.Version` and `RuntimeInformation.OSDescription` continue to return the version returned by [uname](https://www.unix.com/man-page/sunos/2/uname/). For getting Linux distro name and version information, the recommended approach is to read the [`/etc/os-release`](http://man7.org/linux/man-pages/man5/os-release.5.html) file.

### Category

- Core .NET libraries

### Affected APIs

All under the `Microsoft.DotNet.PlatformAbstractions` namespace.

* ApplicationEnvironment.ApplicationBasePath
* HashCodeCombiner
* RuntimeEnvironment.GetRuntimeIdentifier()
* RuntimeEnvironment.OperatingSystemPlatform
* RuntimeEnvironment.RuntimeArchitecture
* RuntimeEnvironment.OperatingSystem
* RuntimeEnvironment.OperatingSystemVersion

---
#### Issue metadata

* Issue type: breaking-change

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #18438.

[Preview link](https://review.docs.microsoft.com/en-us/dotnet/core/compatibility/3.1-5.0?branch=pr-en-us-19010#microsoftdotnetplatformabstractions-package-removed).

### CODE REVIEW

**Comment by gewarren** [comment]
@eerhardt A couple items of note: I could not find the Environment.OSVersion.Version property that was mentioned in #18438. Instead, I referenced System.Environment.OSVersion. Is that correct? Also, most of the PlatformAbstractions APIs I found on the .NET API browser seem to be in a different namespace (Microsoft.Extensions.PlatformAbstractions). Is that the same thing?

File: includes/core-changes/corefx/5.0/platformabstractions-package-removed.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+### Microsoft.DotNet.PlatformAbstractions package removed
+
+No new versions of the [Microsoft.DotNet.PlatformAbstractions NuGet package](https://www.nuget.org/packages/Microsoft.DotNet.PlatformAbstractions/) will be produced, going forward.
```
Comment (tdykstra) [inline includes/core-changes/corefx/5.0/platformabstractions-package-removed.md:3]:
```suggestion
No new versions of the [Microsoft.DotNet.PlatformAbstractions NuGet package](https://www.nuget.org/packages/Microsoft.DotNet.PlatformAbstractions/) will be produced.
```

**Review by tdykstra — APPROVED** [review]
LGTM, one nit suggestion.

File: includes/core-changes/corefx/5.0/platformabstractions-package-removed.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+#### Change description
+
+Previously, new versions of the <xref:Microsoft.DotNet.PlatformAbstractions?displayProperty=fullName> library were produced alongside new versions of .NET Core. Going forward, no new functionality will be added to the library, and no new major versions will be released. However, existing versions of the library will continue to work and be serviced.
+
+The <xref:Microsoft.DotNet.PlatformAbstractions?displayProperty=fullName> library overlaps with APIs that are already established in the System.\* namespaces. Also, some <xref:Microsoft.DotNet.PlatformAbstractions> APIs weren't designed with the same level of scrutiny and long-term supportability as the rest of the System.\* APIs. For example, <xref:Microsoft.DotNet.PlatformAbstractions> uses the `Platform` enumeration to describe the current operating system platform. This enumeration design was explicitly rejected when the <xref:System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform)?displayProperty=nameWithType> API was designed to allow for new platforms and future flexibility.
```
Comment (gewarren) [inline includes/core-changes/corefx/5.0/platformabstractions-package-removed.md:9]:
```suggestion
The <xref:Microsoft.DotNet.PlatformAbstractions?displayProperty=fullName> library overlaps with APIs that are already established in the System.\* namespaces. Also, some <xref:Microsoft.DotNet.PlatformAbstractions> APIs weren't designed with the same level of scrutiny and long-term supportability as the rest of the System.\* APIs. For example, <xref:Microsoft.DotNet.PlatformAbstractions> uses the `Platform` enumeration to describe the current operating system platform. This enumeration design was explicitly rejected when the <xref:System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform)?displayProperty=nameWithType> API was designed, to allow for new platforms and future flexibility.
```

File: includes/core-changes/corefx/5.0/platformabstractions-package-removed.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+#### Version introduced
+
+5.0 Preview 6
+
+#### Recommended action
+
+- You can continue to use older versions of the library if they meet your requirements.
+
+- If the older versions don't meet your requirements, replace usages of the `PlatformAbstractions` APIs with the recommended API, as follows.
```
Comment (gewarren) [inline includes/core-changes/corefx/5.0/platformabstractions-package-removed.md:21]:
```suggestion
- If the older versions don't meet your requirements, replace usages of the `PlatformAbstractions` APIs with the recommended replacements.
```

File: includes/core-changes/corefx/5.0/platformabstractions-package-removed.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  | `PlatformAbstractions` API | Recommended Replacement |
```
Comment (gewarren) [inline includes/core-changes/corefx/5.0/platformabstractions-package-removed.md:23]:
```suggestion
  | `PlatformAbstractions` API | Recommended replacement |
```

**Review by eerhardt — APPROVED** [review]
> I could not find the Environment.OSVersion.Version property that was mentioned in #18438. Instead, I referenced System.Environment.OSVersion. Is that correct?

Yes. `System.Environment.OSVersion` returns a `System.OperatingSystem` object, which has a `Version` property.

> Also, most of the PlatformAbstractions APIs I found on the .NET API browser seem to be in a different namespace (Microsoft.Extensions.PlatformAbstractions). Is that the same thing?

No, unfortunately in the early days of .NET Core we had some package naming confusion, and this was one of them. (You can see a bit of that conversation [here](https://github.com/dotnet/core-setup/pull/231#issuecomment-237607101)). `Microsoft.Extensions.PlatformAbstractions` is a different package, that is also no longer producing new versions.

**Pull request merged**